### PR TITLE
[charts] Remove deprecated `legend` props

### PR DESCRIPTION
**Pull request by alexfauquette**
Part of #13820

This PR is extracted from #15071 I plan to open one PR per set of components, such that the points in the changelog are more informative than "remove breaking changes"

If you prefer another approach feel free to ask

## Changelog

- The `legend` prop got removed. To pass props to the legend, use `slotProps={{ legend: { ... } }}` instead. This can be automatically done with the codemod as long as the `legend` prop does not come from a destructured object.

  ```jsx
  // Supported by the codemod
  <LineChart legend={{ ... }} />

  // Not supported by the codemod
  const props = { legend: { ... }, ... }
  <Line {...props} />
  ```

- The `slots.legend` does not receive the `drawingArea` prop. You can still access your custom legend with the `useDrawingArea()` hook if your custom legend needs it.

### CODE REVIEW

File: packages/x-charts/src/BarChart/BarChart.tsx [hunk 1]
```diff
@@ -80,9 +75,9 @@ export interface BarChartProps
    */
   axisHighlight?: ChartsAxisHighlightProps;
   /**
-   * @deprecated Consider using `slotProps.legend` instead.
+   * If `true`, the legend is not rendered.
    */
-  legend?: ChartsLegendProps;
+  hideLegend?: boolean;
```
Comment (LukasTy) [inline packages/x-charts/src/BarChart/BarChart.tsx:80]: Nitpick: Have you discussed and agreed on this approach?
Couldn't it cause a slight confusion, when all the `legend` props are under `slotProps.legend`, but the `hidden` toggle is top level? 🤔 

P.S. On Pickers we have `tabs` and `toolbar`, which have `hidden` flags under `slotProps`.

P.P.S. Have you considered creating a `ChartsCommonProps` type to reuse `grid`, `axisHighlight`, `hideLegend`, and maybe some other props? 🤔

File: packages/x-codemod/src/v8.0.0/charts/preset-safe/preset-safe.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import { expect } from 'chai';
+import jscodeshift from 'jscodeshift';
+import transform from './index';
+import readFile from '../../../util/readFile';
+
+function read(fileName) {
+  return readFile(path.join(__dirname, fileName));
+}
+
+describe('v8.0.0/pickers', () => {
```
Comment (LukasTy) [inline packages/x-codemod/src/v8.0.0/charts/preset-safe/preset-safe.test.ts:11]:
```suggestion
describe('v8.0.0/charts', () => {
```

File: packages/x-codemod/src/v8.0.0/charts/rename-legend-to-slots-legend/index.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+          }
+
+          const legendProps = elementPath.node.openingElement.attributes?.find(
+            (elementNode) =>
+              elementNode.type === 'JSXAttribute' && elementNode.name.name === 'legend',
+          );
+          const slotProps = elementPath.node.openingElement.attributes?.find(
+            (elementNode) =>
+              elementNode.type === 'JSXAttribute' && elementNode.name.name === 'slotProps',
+          );
```
Comment (LukasTy) [inline packages/x-codemod/src/v8.0.0/charts/rename-legend-to-slots-legend/index.ts:33]: Absolute nitpick: This could be moved after L38.

**Review by LukasTy — APPROVED** [review]
LGTM, nice cleanup. 👍 

Looks ready to merge If the decision on `hideLegend` top level prop is clear. 👌

File: packages/x-charts/src/BarChart/BarChart.tsx (hunk 1 above)
Comment (alexfauquette) [reply packages/x-charts/src/BarChart/BarChart.tsx]: I moved the hideLegend to a dedicated issue to discuss that with José when he come back #15186

**Pull request merged**